### PR TITLE
Benchmark helper for command generation and pretty printing

### PR DESCRIPTION
**Pull request by anijain2305**
A wrapper over the benchmark infrastructure to generate commands, parse the results, and automatically generate graphs.

For this command,

`python benchmarks/helper.py --e2e --modes=training --devices=cuda --suites=torchbench --dtypes=float32 --nightly`

The results look like as follows
~~~
+------+-----------------------------------+------------+-------------+
| dev  |               name                | ts_nvfuser | aot_nvfuser |
+------+-----------------------------------+------------+-------------+
| cuda |    mobilenet_v2_quantized_qat     |    0.0     |   1.7448    |
| cuda |           BERT_pytorch            |    0.0     |   1.4675    |
| cuda |             hf_Albert             |    0.0     |   1.4386    |
| cuda |        mobilenet_v3_large         |   1.079    |    1.427    |
| cuda |         timm_efficientnet         |   1.0983   |   1.4207    |
| cuda |            densenet121            |   1.1109   |   1.4051    |
| cuda |             resnet18              |   1.1008   |   1.3889    |
| cuda |           mobilenet_v2            |   0.9965   |   1.3889    |
| cuda |            mnasnet1_0             |   1.0489   |    1.367    |
| cuda |        shufflenet_v2_x1_0         |   1.0187   |   1.3618    |
| cuda |      timm_vision_transformer      |   0.9522   |    1.353    |
| cuda |          resnext50_32x4d          |   1.115    |   1.3385    |
| cuda |             resnet50              |   0.9943   |   1.2905    |
| cuda |           timm_resnest            |   0.9954   |   1.2903    |
| cuda |              yolov3               |    0.0     |   1.2883    |
| cuda |              hf_GPT2              |    0.0     |   1.2801    |
| cuda |   pytorch_CycleGAN_and_pix2pix    |   1.1076   |   1.2598    |
| cuda |          LearningToPaint          |   1.1291   |   1.2542    |
| cuda |            timm_regnet            |   1.0334   |   1.2356    |
| cuda |            timm_nfnet             |   1.071    |   1.2185    |
| cuda |           pytorch_unet            |   0.9976   |   1.2057    |
| cuda |            timm_vovnet            |   0.994    |   1.1902    |
| cuda |              hf_Bart              |    0.0     |   1.1636    |
| cuda |        Background_Matting         |   1.009    |   1.1551    |
| cuda |           squeezenet1_1           |   1.0532   |   1.1275    |
| cuda |               dcgan               |   1.0746   |   1.1182    |
| cuda |                drq                |    0.0     |   1.1067    |
| cuda |          pytorch_stargan          |   0.9892   |   1.1039    |
| cuda |              hf_Bert              |    0.0     |   1.1019    |
| cuda |           maml_omniglot           |    0.0     |   1.0831    |
| cuda |         timm_efficientdet         |    0.0     |   1.0585    |
| cuda |           hf_Longformer           |    0.0     |   1.0299    |
| cuda |        speech_transformer         |    0.0     |   1.0236    |
| cuda |            hf_BigBird             |    0.0     |    1.02     |
| cuda | attention_is_all_you_need_pytorch |    0.0     |   1.0181    |
| cuda |              demucs               |    0.0     |   0.9995    |
| cuda |              alexnet              |   0.9923   |   0.9981    |
| cuda |               vgg16               |   0.9992   |   0.9969    |
| cuda |          pytorch_struct           |    0.0     |   0.9959    |
| cuda |         soft_actor_critic         |    0.0     |   0.9919    |
| cuda |            Super_SloMo            |   1.004    |   0.9858    |
| cuda |             tacotron2             |    0.0     |    0.985    |
| cuda |            tts_angular            |    0.0     |   0.9798    |
| cuda |          vision_maskrcnn          |    0.0     |   0.9764    |
| cuda |      nvidia_deeprecommender       |    0.0     |   0.9764    |
| cuda |               hf_T5               |    0.0     |    0.887    |
| cuda |           fastNLP_Bert            |    0.0     |   0.8409    |
| cuda |      resnet50_quantized_qat       |    0.0     |     0.0     |
| cuda |           hf_DistilBert           |    0.0     |     0.0     |
| cuda |    detectron2_maskrcnn_r_50_c4    |    0.0     |     0.0     |
| cuda |   detectron2_maskrcnn_r_50_fpn    |    0.0     |     0.0     |
| cuda |               dlrm                |    0.0     |     0.0     |
| cuda |           fambench_dlrm           |    0.0     |     0.0     |
| cuda |  detectron2_fasterrcnn_r_50_fpn   |    0.0     |     0.0     |
| cuda |            hf_Reformer            |    0.0     |     0.0     |
| cuda |  detectron2_fasterrcnn_r_101_c4   |    0.0     |     0.0     |
| cuda |  detectron2_fasterrcnn_r_50_dc5   |    0.0     |     0.0     |
| cuda |   detectron2_fasterrcnn_r_50_c4   |    0.0     |     0.0     |
| cuda |  detectron2_fasterrcnn_r_101_fpn  |    0.0     |     0.0     |
| cuda |  detectron2_fasterrcnn_r_101_dc5  |    0.0     |     0.0     |
| cuda |               moco                |    0.0     |     0.0     |
| cuda |   detectron2_maskrcnn_r_101_fpn   |    0.0     |     0.0     |
| cuda |   detectron2_maskrcnn_r_101_c4    |    0.0     |     0.0     |
+------+-----------------------------------+------------+-------------+
~~~

image

### CODE REVIEW

**Comment by anijain2305** [comment]
cc @ngimel @jansel 

@wconstab Does this help little bit? It is kind of a wrapper, and a place where we can add/modify the latest commands.

**Comment by wconstab** [comment]
Looks pretty good to me, thanks for writing this!

I was confused why these flags
`python benchmarks/helper.py --e2e --modes=training --devices=cuda --suites=torchbench --dtypes=float32 --nightly`
(notably, `--modes=training` --devices=cuda`)
produced outputs for both `ts_nvfuser | aot_nvfuser` and I also wondered how inductor vs nvfuser would be specified.

Initially I had thought of modifying torchbench.py directly so its argparser had these aliases as a one-stop-shop.  But your way might be better, and especially so if you need to call other scripts outside of torchbench.py for HF or TIMM.  

Should we rename?`torchbench.py` still seems like the main script you'd run, and `helper` wouldn't be the obvious thing to try running.  We should make `helper` seem like the main one and the other one seem like a subservient one.

**Comment by ngimel** [comment]
What does `--e2e` flag mean?

**Comment by ngimel** [comment]
>Should we rename?torchbench.py still seems like the main script you'd run

The main script depends on the suite, torchbench.py for torchbench, huggingface.py for huggingface, and soon timm.py for timm.

File: benchmarks/helper.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+        "ts_nnc": "-dcuda --isolate --speedup-ts",
+        "ts_nvfuser": "-dcuda --isolate -n100 --speedup-ts --nvfuser",
+        "trt": "-dcuda --isolate -n100 --speedup-trt",
+        "eager_cudagraphs": "-dcuda --inductor-settings --float32 -n50 --backend=cudagraphs",
+        "nnc_cudagraphs": "-dcuda --inductor-settings --float32 -n50 --backend=cudagraphs_ts --nvfuser",
+        "nvfuser_cudagraphs": "-dcuda --inductor-settings --float32 -n50 --backend=cudagraphs_ts",
+        "inductor_cudagraphs": "-dcuda --inductor-settings --float32 -n50 --inductor",
+        "nightly": ["nvfuser_cudagraphs", "inductor_cudagraphs"],
+    },
+}
```
Comment (anijain2305) [inline benchmarks/helper.py:69]: @wconstab This is the data structure to specify the space for benchmarks.

For "training" and "inference" field, we have (compiler, cmd) pair. There is a special field - nightly (which is a bad name in the hindsight) where you can specify the top2/3 compilers which we are interested in. Thats why we chose the `ts_nvfuser` and `aot_nvfuser`

File: benchmarks/helper.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        "--output-dir", help="Choose the output directory to save the logs"
+    )
+
+    # Choose either generation of commands, pretty parsing or e2e runs
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument("--commands", action="store_true", help="Generate commands")
+    group.add_argument("--parse", action="store_true", help="Parse the files")
+    group.add_argument(
+        "--e2e", action="store_true", help="Generate commands, run and parse the files"
+    )
```
Comment (anijain2305) [inline benchmarks/helper.py:97]: @ngimel This is the mutually exclusive set of options.

1) Commands - Just generate the commands and save them in run.sh
2) Parse - Parse the already generated logs, and pretty print/graphs etc
3) e2e - (which in hindsight is a hideous name) generates command, runs the tests, and parses the logs as well.

Comment (wconstab) [reply benchmarks/helper.py:97]: maybe make these changes?
  commands -> print_run_commands
  parse -> visualize_logs
  e2e -> run
  
and should 'run' be the default, such that 'required=False' for this group?

File: benchmarks/helper.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+}
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--filter", "-k", action="append", help="filter benchmarks with regexp"
+    )
+    parser.add_argument("--devices", action="append", help="cpu or cuda")
+    parser.add_argument("--modes", action="append", help="inference or training")
```
Comment (wconstab) [inline benchmarks/helper.py:78]: maybe cut out 'modes' arg,

add --training and --inference args which are both append actions,  

have the default for training = ["ts_nvfuser", "aot_nvfuser"]
get rid of the 'nightly' flag altogether since now it is just the 'default' for --training

and it could be an error if you specify both --training and --inference or it could work to specify both, depending how hard that is to implement?

Comment (anijain2305) [reply benchmarks/helper.py:78]: Perfect. This is much better. Thanks!

**Comment by anijain2305** [comment]
@wconstab Another round of review. I had to add `--inference` and `--training` as bool, and assist with a `--compliers` append action arg. It gets ugly to have a default for append action args. 

So, the pretty print and graphs in the OP would be a result of this command

~~~
python benchmarks/helper.py --suites=torchbench --training
~~~

Any better name for the `helper.py`? 

Currently, in the directory, we have obvious torchbench.py and huggingface.py, both of which depend on common.py.

File: benchmarks/helper.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+rcParams.update({"figure.autolayout": True})
+plt.rc("axes", axisbelow=True)
+
+DEFAULT_OUTPUT_DIR = "benchmark_logs"
+
+
+TABLE = {
+    "training": {
+        "ts_nnc": "--training --speedup-ts --use-eval-mode --isolate",
```
Comment (wconstab) [inline benchmarks/helper.py:43]: i'm curious- what does ts_nnc do? does it just compile forward graph and not compile backward?

or it uses torchscript autodiff and relies on that for bwd?

File: benchmarks/helper.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+
+    # Graph
+    labels = df.columns.values.tolist()
+    labels = labels[2:]
+    df.plot(
+        x="name",
+        y=labels,
+        kind="bar",
+        title=title,
+        ylabel="Speedeup over eager",
```
Comment (wconstab) [inline benchmarks/helper.py:177]: typo speedup

File: benchmarks/helper.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+
+    output_dir = args.output_dir if args.output_dir is not None else DEFAULT_OUTPUT_DIR
+
+    if args.print_run_commands:
+        generate_commands(args, dtypes, suites, devices, compilers, output_dir)
+    elif args.visualize_logs:
+        parse_logs(args, dtypes, suites, devices, compilers, output_dir)
+    elif args.run:
+        generate_commands(args, dtypes, suites, devices, compilers, output_dir)
+        os.system("bash run.sh")
```
Comment (wconstab) [inline benchmarks/helper.py:240]: does this do enough to flag / catch if run.sh errs?

File: benchmarks/helper.py (hunk 4 above)
Comment (anijain2305) [reply benchmarks/helper.py:43]: Latter. Torchscript autodiff. Torchscript is used to capture the fwd and bwd graphs.
NNC or NVFuser are the backend compilers once the graphs are captured by torchscript.

File: benchmarks/helper.py (hunk 6 above)
Comment (anijain2305) [reply benchmarks/helper.py:240]: Ah, missed that. Thanks!

**Review by wconstab — APPROVED** [review]
thanks Animesh. This will be a big help to everyone.

I left a few nits.

I also think it'd be better to rename helper.py to something more indicative that it is the main runner.

run.py?
runner.py?
main.py?

If you do change it, make sure to update the comments in your docstrings.

File: benchmarks/helper.py (hunk 4 above)
Comment (anijain2305) [reply benchmarks/helper.py:43]: This does not use TorchDynamo to extract the graphs though. Hmm, I actually meant to use that.

I will add an option in common.py and correspondingly update this file in next PR.
